### ISSUE TITLE
Sftp does not do SFTP

### ISSUE THREAD

**Issue by Ten0** [openssh-rust/openssh#15]
Instead, it is using shell commands.
This makes the tool incompatible with SFTP-only servers.

It's unclear when looking at the documentation. It probably shouldn't be called SFTP.

**Comment by jonhoo** [comment]
Yes, that is correct. I wrote the API this way because I _want_ it to do SFTP, but did not have the time to build that out at the time. The API as written should allow an SFTP implementation to be swapped in for the existing ssh-based one, and I'd be very much in favor of doing so! If this is a use-case you have (specifically needing SFTP), would you be able to give a PR a try?

**Comment by Ten0** [comment]
Unfortunately no, with that + the additional async complexity, + the fact it's not able to list files, +not able to connect with PKeys in memory, etc, I think going back to patching up the [broken](https://github.com/alexcrichton/ssh2-rs/issues/195) `ssh2` lib will be a better approach for my use case.

**Comment by jonhoo** [comment]
Okay, fair enough :) Should you find it too broken in the future, I'm happy to help out with PRs!

**Comment by NobodyXu** [comment]
I've been working on implementing a [SFTP client] in pure rust that can be used concurrently.

Right now I am implementing [concurrent_arena] for it since I believe using `Mutex<Arena>` isn't scalable.

If you are interested, you can help me with adding more tests to [SFTP client] or [concurrent_arena].

[SFTP client]: https://github.com/NobodyXu/openssh-sftp-client
[concurrent_arena]: https://github.com/NobodyXu/concurrent_arena

**Comment by NobodyXu** [comment]
There's now a [data race](https://github.com/NobodyXu/concurrent_arena/issues/1) in [`concurrent_arena`](https://github.com/NobodyXu/concurrent_arena) reported by rust's experimental thread sanitizer, however the report is quite confusing for me and I have trouble understanding the root cause for this error.

@jonhoo If you have time, can you please help me with this?

**Comment by jonhoo** [comment]
Hmm, I wonder if you could just use the excellent [`sharded-slab`](https://crates.io/crates/sharded-slab) for this instead of rolling your own?

**Comment by NobodyXu** [comment]
Sftp protocol uses a `u32` request_id, that is why I cannot use `shared-slab` as there is no guarantee on its return value is less than `u32::MAX`.

**Comment by NobodyXu** [comment]
Perhaps I can open an issue in [shared-slab](https://crates.io/crates/sharded-slab) to see if there is anyway to convert the `u64` slot id to a unique `u32` slot id.

**Comment by NobodyXu** [comment]
It seems that the [data race](https://github.com/NobodyXu/concurrent_arena/issues/1) is just a false positive produced by the thread sanitizer.

And I've updated this crate to use [`arc-swap`](https://github.com/vorner/arc-swap) instead.

So there is no blocking for the development of [`openssh-sftp-client`](https://github.com/NobodyXu/openssh-sftp-client).

**Comment by NobodyXu** [comment]
[openssh-sftp-client](https://github.com/NobodyXu/openssh-sftp-client) now can create new file/open existing file, read/write to it and close it.

These operations are already tested by calling `sftp-server` directly in the unit tests.

**Comment by NobodyXu** [comment]
I have added unit tests for all basic operations of sftp protocol v3.

I will be working on adding support for openssh extensions.

**Comment by NobodyXu** [comment]
I've added and tested the following extensions:
 - limits                                   
 - expand path                              
 - fsync                                    
 - hardlink                                 
 - posix rename

**Comment by NobodyXu** [comment]
Here is the [doc](https://docs.rs/openssh-sftp-client/latest/openssh_sftp_client/index.html) for [`openssh-sftp-client`].

@jonhoo I been wondering how to add true sftp support using [`openssh-sftp-client`] to `openssh-rs`.

Shall I add a new feature and a new API like `Session::real_sftp`?

Or maybe just replace the old "sftp" implementation since it is not implemented using sftp and it is actually quite confusing to users of this crate.

[`openssh-sftp-client`]: https://crates.io/crates/openssh-sftp-client

**Comment by Ten0** [comment]
I think you should replace the old sftp implementation because that's what it was initially named for.
(if you want to keep the existing functions they should probably be named scp)

**Comment by NobodyXu** [comment]
> I think you should replace the old sftp implementation because [that's what it was initially named for](https://github.com/jonhoo/openssh-rs/issues/15#issuecomment-671385023). (if you want to keep the existing functions they should probably be named scp)

LGTM.

@jonhoo Is it OK to completely replace the old sftp implementation completely and adding some new functions?

**Comment by jonhoo** [comment]
Absolutely!

**Comment by NobodyXu** [comment]
> Absolutely!

Great, I will create a new PR once #26 and #28 is merged.

**Comment by NobodyXu** [comment]
Update on [`openssh-sftp-client`](https://crates.io/crates/openssh-sftp-client):

v0.3.0 is released, adding:
 - Write Buffer to [`WriteEnd`];
 - Read-ahead buffer to [`ReadEnd`];
 - Thanks to adding write buffer to [`WriteEnd`], now most of the methods in [`WriteEnd`] is cancel safe.
   
   In fact, most of the methods are not `async` anymore and relies on [`WriteEnd::flush`] to flush the data, which is also
   cancel safe, unless used with [`WriteEnd::send_write_request_direct`] or [`WriteEnd::send_write_request_direct_vectored`].

[`WriteEnd`]: https://docs.rs/openssh-sftp-client/0.3.0/openssh_sftp_client/struct.WriteEnd.html
[`ReadEnd`]: https://docs.rs/openssh-sftp-client/0.3.0/openssh_sftp_client/struct.ReadEnd.html
[`WriteEnd::flush`]: https://docs.rs/openssh-sftp-client/0.3.0/openssh_sftp_client/struct.WriteEnd.html#method.flush
[`WriteEnd::send_write_request_direct`]: https://docs.rs/openssh-sftp-client/0.3.0/openssh_sftp_client/struct.WriteEnd.html#method.send_write_request_direct
[`WriteEnd::send_write_request_direct_vectored`]: https://docs.rs/openssh-sftp-client/0.3.0/openssh_sftp_client/struct.WriteEnd.html#method.send_write_request_direct_vectored

**Comment by NobodyXu** [comment]
@Ten0 The draft PR is ready for review.

**Comment by NobodyXu** [comment]
@Ten0 [`openssh-0.9.1`] has released, with support for sftp subsystem, which can be used with [`openssh-sftp-client`].

[`openssh-0.9.1`]: https://docs.rs/openssh/0.9.1/openssh/
[`openssh-sftp-client`]: https://docs.rs/openssh-sftp-client/latest/openssh_sftp_client/
